src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,1675 @@
+    auto const sleB = psb.peek(keylet::bridge(bridgeSpec));
+    auto const sleCID = psb.peek(keylet::xChainClaimID(bridgeSpec, claimID));
+
+    if (!(sleB && sleCID && sleAcc))
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    AccountID const thisDoor = (*sleB)[sfAccount];
+
+    STXChainBridge::ChainType dstChain = STXChainBridge::ChainType::locking;
+    {
+        if (thisDoor == bridgeSpec.lockingChainDoor())
+            dstChain = STXChainBridge::ChainType::locking;
+        else if (thisDoor == bridgeSpec.issuingChainDoor())
+            dstChain = STXChainBridge::ChainType::issuing;
+        else
+            return tecINTERNAL;  // LCOV_EXCL_LINE
+    }
+    STXChainBridge::ChainType const srcChain =
+        STXChainBridge::otherChain(dstChain);
+
+    auto const sendingAmount = [&]() -> STAmount {
+        STAmount r(thisChainAmount);
+        r.setIssue(bridgeSpec.issue(srcChain));
+        return r;
+    }();
+
+    auto const [signersList, quorum, slTer] =
+        getSignersListAndQuorum(ctx_.view(), *sleB, ctx_.journal);
+
+    if (!isTesSuccess(slTer))
+        return slTer;  // LCOV_EXCL_LINE
+
+    XChainClaimAttestations curAtts{
+        sleCID->getFieldArray(sfXChainClaimAttestations)};
+
+    auto claimR = curAtts.onClaim(
+        sendingAmount,
+        /*wasLockingChainSend*/ srcChain == STXChainBridge::ChainType::locking,
+        quorum,
+        signersList);
+    if (!claimR.has_value())
+        return claimR.error();
+
+    auto const& rewardAccounts = claimR.value();
+    auto const& rewardPoolSrc = (*sleCID)[sfAccount];
+
+    std::optional<std::uint32_t> dstTag = ctx_.tx[~sfDestinationTag];
+
+    auto const r = finalizeClaimHelper(
+        psb,
+        bridgeSpec,
+        dst,
+        dstTag,
+        /*claimOwner*/ account,
+        sendingAmount,
+        rewardPoolSrc,
+        (*sleCID)[sfSignatureReward],
+        rewardAccounts,
+        srcChain,
+        sleCID,
+        OnTransferFail::keepClaim,
+        ctx_.journal);
+    if (!isTesSuccess(r))
+        return r;
+
+    psb.apply(ctx_.rawView());
+
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TxConsequences
+XChainCommit::makeTxConsequences(PreflightContext const& ctx)
+{
+    auto const maxSpend = [&] {
+        auto const amount = ctx.tx[sfAmount];
+        if (amount.native() && amount.signum() > 0)
+            return amount.xrp();
+        return XRPAmount{beast::zero};
+    }();
+
+    return TxConsequences{ctx.tx, maxSpend};
+}
+
+NotTEC
+XChainCommit::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureXChainBridge))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;  // LCOV_EXCL_LINE
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const amount = ctx.tx[sfAmount];
+
+    if (amount.signum() <= 0 || !isLegalNet(amount))
```

Comment:
Shouldn't this have the same check that is done in `XChainClaim`?
```cpp
if (amount.signum() <= 0 ||
  (amount.issue() != bridgeSpec.lockingChainIssue() &&
   amount.issue() != bridgeSpec.issuingChainIssue()))
{
    return temBAD_AMOUNT;
}
```

Reply:
We do a more complete check in `preclaim`. I don't think we need the extra check here.

Reply:
We do a more complete check in `preclaim` so the check isn't needed here. But it is better to catch errors in `preflight` if we can. Added the check.